### PR TITLE
feat: Cloudflare R2 provider

### PR DESCRIPTION
**Pull request by l4r-s**
Talked about this already on Twitter.
I think I could fix the bug I  mentioned there.

Let me know if there is something that needs work or is unclear :)

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -53,6 +53,19 @@ export type ProviderConfig = {
     /* An optional function to generate the bucket asset key. */
     generateAssetKey?: (filePathOrURL: string, folder: string) => string;
   };
+
+  'cloudflare-r2'?: {
+    endpoint: string;
+    accountId: string;
+    bucket?: string;
+    bucketUrlPublic?: string;
```
Comment (luwes) [inline src/config.ts:61]: is there any way `bucketUrlPublic` can be constructed from endpoint, accountId and bucket?

File: src/config.ts [hunk 2]
```diff
@@ -53,6 +53,19 @@ export type ProviderConfig = {
...
+    accessKeyId?: string;
+    secretAccessKey?: string;
+    apiToken?: string;
+    jurisdiction?: string;
```
Comment (luwes) [inline src/config.ts:65]: let's remove jurisdiction and use the endpoint config to keep it more uniform with other providers

File: src/config.ts [hunk 3]
```diff
@@ -53,6 +53,19 @@ export type ProviderConfig = {
     /* An optional function to generate the bucket asset key. */
     generateAssetKey?: (filePathOrURL: string, folder: string) => string;
   };
+
+  'cloudflare-r2'?: {
+    endpoint: string;
+    accountId: string;
```
Comment (luwes) [inline src/config.ts:59]: remove accountId too, it's part of endpoint as well. we can extract it from the endpoint if needed.

File: src/config.ts (hunk 2 above)
Comment (l4r-s) [reply src/config.ts]: 👍 changed it

File: src/config.ts (hunk 3 above)
Comment (l4r-s) [reply src/config.ts]: 👍 changed it. Will now take it from the endpoint url

File: src/config.ts (hunk 1 above)
Comment (l4r-s) [reply src/config.ts:60]: Without a Cloudflare API Token (R2_CF_API_TOKEN) to my knowledge no.

There are two ways for public access:
- r2.dev subdomain
- custom domain
Both must be enabled first on a bucket level.

If a Cloudflare API Token is given to the provider the r2.dev subdomain will be enabled when creating the bucket (if none is given). Making it usable instantly without manual interaction.

**Comment by l4r-s** [comment]
sorry for my late comments. Let me know if there is anything else :)

**Pull request merged**